File: backend/src/models/edition.py
"""Edition-related enums and models for the Editions feature.

These enums define the different edition types, fidelity levels,
and coverage strengths used throughout the system.

Models include:
- SegmentRef: Reference to a transcript segment with canonical_hash for offset drift protection
- Theme: A theme/chapter for Ideas Edition
- SpeakerRole: Role of speaker in transcript (host, guest, caller, clip, unclear)
- SpeakerRef: Reference to a speaker with typed role for whitelist-based quote generation
- TranscriptPair: Both transcript forms needed for whitelist building
- WhitelistQuote: A validated quote that can be used in generation
"""

from enum import Enum
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class Edition(str, Enum):
    """Type of edition to generate.

    QA: Question-and-answer format preserving interview structure
    IDEAS: Key ideas extraction format
    """

    QA = "qa"
    IDEAS = "ideas"


class Fidelity(str, Enum):
    """Fidelity level for content generation.

    FAITHFUL: Maintains meaning while allowing minor rephrasing
    VERBATIM: Exact word-for-word preservation
    """

    FAITHFUL = "faithful"
    VERBATIM = "verbatim"


class Coverage(str, Enum):
    """Coverage strength indicating how well content is supported.

    STRONG: Content is well-supported by evidence
    MEDIUM: Content has moderate support
    WEAK: Content has minimal support
    """

    STRONG = "strong"
    MEDIUM = "medium"
    WEAK = "weak"


class CoverageLevel(str, Enum):
    """Coverage strength for quote availability.

    STRONG: >= 5 usable quotes, >= 50 words/claim
    MEDIUM: >= 3 usable quotes, >= 30 words/claim
    WEAK: Below MEDIUM thresholds
    """

    STRONG = "strong"
    MEDIUM = "medium"
    WEAK = "weak"


class ChapterCoverage(BaseModel):
    """Coverage metrics for a chapter."""

    model_config = ConfigDict(extra="forbid")

    chapter_index: int = Field(ge=0)
    level: CoverageLevel
    usable_quotes: int = Field(ge=0)
    quote_words_per_claim: float = Field(ge=0)
    quotes_per_claim: float = Field(ge=0)
    target_words: int = Field(ge=0)
    generation_mode: Literal["normal", "thin", "excerpt_only"] = Field(
        description="Generation mode based on coverage level"
    )


class SpeakerRole(str, Enum):
    """Role of speaker in transcript.

    HOST: The host/interviewer of the podcast/webinar
    GUEST: A guest being interviewed or featured
    CALLER: A caller or audience member
    CLIP: Audio/video clip from another source
    UNCLEAR: Speaker role cannot be determined
    """

    HOST = "host"
    GUEST = "guest"
    CALLER = "caller"
    CLIP = "clip"
    UNCLEAR = "unclear"


class SpeakerRef(BaseModel):
    """Reference to a speaker with typed role.

    Used for whitelist-based quote generation to filter quotes by speaker role.
    """

    model_config = ConfigDict(extra="forbid")

    speaker_id: str = Field(description="Canonical stable ID (e.g., 'david_deutsch')")
    speaker_name: str = Field(description="Display name (e.g., 'David Deutsch')")
    speaker_role: SpeakerRole = Field(description="Role for filtering")


class TranscriptPair(BaseModel):
    """Both transcript forms needed for whitelist building."""

    model_config = ConfigDict(extra="forbid")

    raw: str = Field(description="Original transcript (for quote_text extraction)")
    canonical: str = Field(description="Normalized (for matching)")


class WhitelistQuote(BaseModel):
    """A validated quote that can be used in generation."""

    model_config = ConfigDict(extra="forbid")

    quote_id: str = Field(
        min_length=16,
        max_length=16,
        pattern=r"^[a-f0-9]{16}$",
        description="Stable ID: sha256(speaker_id|quote_canonical)[:16]"
    )
    quote_text: str = Field(description="EXACT from raw transcript (for output)")
    quote_canonical: str = Field(description="Casefolded/normalized (for matching only)")
    speaker: SpeakerRef
    source_evidence_ids: list[str] = Field(default_factory=list)
    chapter_indices: list[int] = Field(default_factory=list)
    match_spans: list[tuple[int, int]] = Field(default_factory=list)


class SegmentRef(BaseModel):
    """Reference to a transcript segment.

    CRITICAL: canonical_hash stores the SHA256 hash of the canonical transcript
    these offsets reference. This prevents offset drift if the transcript changes.
    """

    model_config = ConfigDict(extra="forbid")

    start_offset: Annotated[int, Field(ge=0, description="Start character offset, must be >= 0")]
    end_offset: Annotated[int, Field(ge=0, description="End character offset, must be >= 0")]
    token_count: Annotated[int, Field(ge=0, description="Actual token count, NOT from preview")]
    text_preview: str = Field(description="First ~100 chars for display only")
    canonical_hash: str = Field(
        description="SHA256 hash of canonical transcript (REQUIRED for offset validity)"
    )


class Theme(BaseModel):
    """A theme/chapter for Ideas Edition."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(description="Unique theme identifier")
    title: str = Field(description="Theme title")
    one_liner: str = Field(description="Brief description of the theme")
    keywords: list[str] = Field(description="Keywords associated with this theme")
    coverage: Coverage = Field(description="Coverage strength of this theme")
    supporting_segments: list[SegmentRef] = Field(
        description="Transcript segments supporting this theme"
    )
    include_in_generation: bool = Field(
        default=True,
        description="Whether to include this theme in generation"
    )


def get_recommended_edition(webinar_type: str) -> Edition:
    """Get recommended edition based on webinar type.

    - Presentations and tutorials → Ideas Edition (thematic chapters)
    - Interviews → Q&A Edition (speaker-labeled dialogue)

    Args:
        webinar_type: The webinar type string value

    Returns:
        The recommended Edition enum value
    """
    if webinar_type == "interview":
        return Edition.QA
    # standard_presentation, training_tutorial, or any other type
    return Edition.IDEAS

File: frontend/src/types/edition.ts
/**
 * Edition types for output format selection.
 *
 * Two editions available:
 * - Q&A Edition: Faithful interview format with speaker labels
 * - Ideas Edition: Thematic chapters with synthesized prose
 */

import type { WebinarType } from './project'

/** Output edition type */
export type Edition = 'qa' | 'ideas'

/** Fidelity level for Q&A Edition */
export type Fidelity = 'faithful' | 'verbatim'

/** Theme coverage strength */
export type Coverage = 'strong' | 'medium' | 'weak'

/**
 * Reference to a transcript segment.
 *
 * CRITICAL: canonical_hash stores the SHA256 hash of the canonical transcript
 * these offsets reference. This prevents offset drift if the transcript changes.
 */
export interface SegmentRef {
  start_offset: number
  end_offset: number
  token_count: number
  text_preview: string
  /** SHA256 hash of canonical transcript (REQUIRED for offset validity) */
  canonical_hash: string
}

/** Theme/chapter for Ideas Edition */
export interface Theme {
  id: string
  title: string
  one_liner: string
  keywords: string[]
  coverage: Coverage
  supporting_segments: SegmentRef[]
  include_in_generation: boolean
}

/** Theme proposal job status */
export type ThemeJobStatus = 'queued' | 'processing' | 'completed' | 'failed' | 'cancelled'

/** Theme proposal job */
export interface ThemeJob {
  job_id: string
  project_id: string
  status: ThemeJobStatus
  themes: Theme[]
  error: string | null
}

// UI Labels and Descriptions

export const EDITION_LABELS: Record<Edition, string> = {
  qa: 'Q&A Edition',
  ideas: 'Ideas Edition',
}

export const EDITION_DESCRIPTIONS: Record<Edition, string> = {
  qa: 'Faithful interview format with speaker labels',
  ideas: 'Thematic chapters with synthesized prose',
}

export const FIDELITY_LABELS: Record<Fidelity, string> = {
  faithful: 'Faithful (cleaned)',
  verbatim: 'Verbatim (strict)',
}

export const FIDELITY_DESCRIPTIONS: Record<Fidelity, string> = {
  faithful: 'Cleaned transcript with light editing for readability',
  verbatim: 'Exact word-for-word preservation of speaker words',
}

export const COVERAGE_LABELS: Record<Coverage, string> = {
  strong: 'Strong',
  medium: 'Medium',
  weak: 'Weak',
}

export const COVERAGE_COLORS: Record<Coverage, string> = {
  strong: 'text-green-600 bg-green-100',
  medium: 'text-yellow-600 bg-yellow-100',
  weak: 'text-red-600 bg-red-100',
}

/** Default fidelity for Q&A Edition */
export const DEFAULT_FIDELITY: Fidelity = 'faithful'

/**
 * Get recommended edition based on webinar type.
 *
 * - Presentations and tutorials → Ideas Edition (thematic chapters)
 * - Interviews → Q&A Edition (speaker-labeled dialogue)
 */
export function getRecommendedEdition(webinarType: WebinarType): Edition {
  switch (webinarType) {
    case 'interview':
      return 'qa'
    case 'standard_presentation':
    case 'training_tutorial':
    default:
      return 'ideas'
  }
}

export const EDITION_RECOMMENDATION_REASONS: Record<Edition, string> = {
  qa: 'Best for interview-style content with multiple speakers',
  ideas: 'Best for presentations and tutorials organized by topics',
}
